Muon - Enabling Multi-threading in the new GUI

TWJubb opened the pull request:
closes #23296 

Enabling multi-threading for Muon GUI
-------------------------------------

There are several places in the new muon GUI where multi-threading may be beneficial (e.g. loading, fitting, anything involving updating multiple runs).

This PR creates the code to enable this behaviour and also adds several benefits to the user:

- Progress tracking, a `QProgressBar` can (and will) be easily implemented.
- User controlled number of threads.
- Callbacks for handling exceptions (and other events) to implement live warning messages to the user whilst the code is executing in the background.
- Clean cancellation of threads is possible at any point.

To use the multi-threading you only need:

- A callable, with lists of each of its input arguments (see example below)
- An instance of `WorkerManager()`

The `WorkerManager` will split the lists of arguments into `n` equally sized chunks (where `n` is the user supplied number of threads to use) and give one chunk to each `Worker` (single thread). The results are stored in a dictionary which contains key:value pairs where the value of the inputs and the results are held.

All exceptions in the evaluation of the function are caught and propagated out of the thread via a callback (rather than  being re-raised); allowing the user to handle exceptions any way they want without exceptions being able to cause a crash.

Optional callbacks can be given to the `WorkerManager` to handle the threads finishing, raising exceptions or updating progress (for the progress bar).

This code has already been implemented in a widget #20943, and PR's for this will follow this one.

To test
-------

I have provided some example test code below. These classes are intended to be flexible and so trying to cause unexpected behaviour is the best way to test.

The threading works for functions with arbitrary inputs, outputs and exception raising, so create a function and try varying each of these behaviours and then attempt to successfully produce results across multiple threads.

Please also comment on : 

- Possible ways to make the code more efficient.
- Readability (as this code can be used by any python GUI's)
- Test coverage/test cases.

Example test code
------------------

Create a function with arbitrary arguments, returning anything, and with possible exception raising, e.g.

```Python
import time

def test_function(x,y):
    time.sleep(1)
    return x / y
```

Now create a `WorkerManager` object and give it lists in place of the usual arguments to the function (it is the client code responsibility to ensure these lists are of equal length)

```Python
manager = WorkerManager(test_function, x=[10,10,10,10,10], y=[1,2,3,4,0])
```

In order to execute correctly, the threads need to execute within an event loop, the simple class below creates an eventloop for the duration of the execution of the threads (there are variations of this class within the unit tests to extend its behaviour to, for example, cancel the threads after a certain amount of time and check nothing bad happens as a result)

```Python
class Runner:
    QT_APP = QtWidgets.QApplication([])

    def __init__(self, thread):
        self._thread = thread
        self._thread.finished.connect(self.finished)
        self._thread.cancelled.connect(self.finished)
        QtCore.QTimer.singleShot(5000, self._thread.clear)

        self.QT_APP.exec_()

    def finished(self):
        self.QT_APP.processEvents()
        self.QT_APP.exit(0)
```

Now the code can be executed as

```Python
manager.start()
Runner(manager)

# All code here will be executed once the threads have finished
```

The results are stored within the `WorkerManager` as `manager.results`.

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

TWJubb committed:
Refs 23296 Initial commit, code + tests.

TWJubb committed:
Refs #23296 Fix flake-8 warnings.

TWJubb commented:
I should mention that this code was benchmarked with the `Load` algorithm and some Muon data. With 10, 50 files the multi-threaded code is around 2x faster with 4 vs 1 threads; so the overhead is not negligible but the improvement is significant.

martyngigg commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+    equally across a specified number of threads.
+
+    Example : for a function f(x,y) where x and y are ints, the WorkerManager allows for evaluation of
+    f(x=[1,2,3,4,5],y=[1,2,3,4,5]) by splitting the lists across the given number of threads. So for two threads the
```
I like the flexibility of allowing arbitrary callables here. The syntax for the multi-argument functions feels a little clunky though as it looks like you have to split the lists up manually?
Could we just require the caller to pass the arguments as lists of tuples, for example: `f(x,y)` with args `((1,1), (2,2), (3,3), (4,4), (5,5))` naturally keeps the arguments together and I think makes it clearer what is being evaluated?

TWJubb replied:
At the moment I have limited arguments to keyword args, and I think that would prevent me doing this. However, I can add support for `*args*` as well, but it will make the `run` method of `Worker` more involved

martyngigg commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+        # maintain a list of running threads to allow for cancelling
+        self._threads = []
+
+        self._results = {key: [] for key in self.kwarg_list.keys() + ['results']}
```
Can we call `self._clear_results` here instead of duplicating the logic over the next 3 lines?

martyngigg commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+        worker.signals.cancelled.connect(self.on_thread_cancelled)
+
+    def on_thread_start(self):
+        pass
```
I think this and the next function could benefit from a comment that they are intentionally left empty.

martyngigg commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+
+    def _clear_results(self):
+        self._progress = 0.0
+        self._results = {key: [] for key in self.kwarg_list.keys() + ['results']}
```
Can you help me understand this? Is it storing the inputs and then an extra field for the results? 

Is there a need to store the results at all? Can we just provide them in an the optional callback given by the caller?

TWJubb replied:
Yes, the results are stored alongside all the inputs used to generate them as the threads will be asynchronously writing to `_results` so I need to be able to identify which result came from which evaluation.

I feel there is a need to store the results so that they can all be accessed at once after all the threads are finished; if you wanted to do that via a callback you would still need to keep the results somewhere no?

martyngigg commented on scripts/test/Muon/threading_manager_test.py:
```diff
@@ -0,0 +1,303 @@
+else:
+    import mock
+
+from Muon.GUI.Common.threading_manager import *
```
Can we avoid `*` imports - they are bad practice so it would be good to avoid new ones.

martyngigg reviewed (commented):
This is generally good and is well coded. I just have a few comments and questions that are inline.

TWJubb commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+        self.terminate()
+
+
+class WorkerManager(QtWidgets.QWidget):
```
Yep, I'll fix that

TWJubb commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+    finished = Signal()  # used for unit tests
+    cancelled = Signal()  # used for unit tests
+
+    def __init__(self, fn, num_threads,
```
Yes, and default of `1` is probably sensible so that multiple threads are only created if specifically requested and the default behaviour is just to start a single "background" thread.

ewancook commented on scripts/test/Muon/threading_manager_test.py:
```diff
@@ -0,0 +1,303 @@
+if sys.version_info.major == 3:
+    from unittest import mock
+else:
+    import mock
```
```py
try:
    from unittest import mock
except ImportError:
    import mock
```
would be more [pythonic](https://docs.python.org/3.6/glossary.html#term-eafp)

ewancook commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+    def _clear_results(self):
+        self._progress = 0.0
+        self._results = {key: [] for key in self.kwarg_list.keys() + ['results']}
+        self._failed_results = {key: [] for key in self.kwarg_list.keys()}
```
```py
self.kwarg_list.keys()
```
returns a list every time it's called - assigning it to a variable then using this in the above two lines would be more efficient

ewancook replied:
Martyn's [comment](https://github.com/mantidproject/mantid/pull/23297/files#r211554593) addresses the other location in which this would've been the case

ewancook commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+                failed_results_exceptions += [e]
+                continue
+            results += [result]
+            progress_callback.emit(100 / num_evals)
```
```py
results += [result]
progress_callback.emit(100 / num_evals)
```
Could be moved inside the `try` block, thus removing the need for `continue`

Although it's a matter of style (and opinionated!), I think `.extend([e])` is more readable than `+= [e]` (and this occurs throughout the code).
`.append(e)` would probably be the most readable choice because you're only adding one element to the list at a time.

ewancook commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+                self.connect_worker_signals(worker)
+                self._threads += [worker]
+                if self.verbose:
+                    print("\tStarting thread " + str(i + 1) + " with arg : ", arg)
```
```py
print("\tStarting thread {} with arg: {}".format(str(i + 1), arg))
```
Is the preferred way to format strings

ewancook commented on scripts/Muon/GUI/Common/threading_manager.py:
```diff
@@ -0,0 +1,282 @@
+
+    def on_thread_complete(self):
+        if self.verbose:
+            print("\tTHREAD " + str(self._thread_count) + " COMPLETE!\n")
```
See string formatting [comment](https://github.com/mantidproject/mantid/pull/23297/files#r211580982)

ewancook commented on scripts/test/Muon/threading_manager_test.py:
```diff
@@ -0,0 +1,303 @@
+    if lst == raise_on:
+        raise ValueError
+    else:
+        return 2 * lst
```
`else` is redundant here, but quite happy for this to be left in for clarity.

ewancook commented on scripts/test/Muon/threading_manager_test.py:
```diff
@@ -0,0 +1,303 @@
+        self.Runner(worker)
+
+        self.assertEqual(worker_result.call_count, 0)
+        self.assertEqual(worker_error.call_count, 1)
```
`worker_result`, `worker_error`, `worker_progress`, `worker_started` and `worker_finished` are used/shared throughout the tests. I'd move these to `setUp()` and make them elements of `self` to clear up the test functions.

Consider creating a function that accepts `Worker(...)` and connects would-be `self.worker.signals...` , calls `worker.start()` and calls `self.Runner(worker)` as these are common to all tests. This could then be called before the asserts.

ewancook replied:
All signals could be connected in such a function, and only the outputs required would be checked in the `self.assertEquals`

ewancook commented on scripts/test/Muon/threading_manager_test.py:
```diff
@@ -0,0 +1,303 @@
+        progress_call_args = [i[0][0] for i in manager_progress.call_args_list]
+        self.assertEqual(progress_call_args[0], 0.0)
+        self.assertEqual(progress_call_args[-1], 100.0)
+        self.assertEqual(sorted(progress_call_args), progress_call_args)
```
similar to [this](https://github.com/mantidproject/mantid/pull/23297/files#r211585978)

ewancook reviewed (changes requested):
As far as I can see, the functionality and code quality (especially testing) looks good - I've suggested a few changes, with details included on whether they're preferred or not: a lot of them are stylistic choices.

However, the tests don't run because they're not in `scripts/test/Muon/CMakeLists.txt`. I got timeout errors from cmake when I tried to run them individually.

TWJubb committed:
Refs #23296 Change how threads are started.

TWJubb commented:
I have updated the way that workers and threads are handled to make it more stable,

see e.g.

https://stackoverflow.com/questions/20324804/how-to-use-qthread-correctly-in-pyqt-with-movetothread

I have also remove the "cancel" functionality of a single thread, since this is redundant in light of the above discussion. Instead, only the manager can be cancelled, and even then it must wait for its threads to finish execution (thus the cancel functionality behaves more like "cancel at the next appropriate time").

This is the only clean way to cancel `QThreads` that I can find. Using `.terminate()` leads to a lot of crashes and instability so I have removed it.

TWJubb committed:
Refs #23296 Add tests to CMake

TWJubb committed:
Refs #23296 Fix compatibility issues.

TWJubb committed:
Refs #23296 Fix flake8 warnings.

TWJubb committed:
Refs #23296 fix imports.
